test/provenance/provenance_query_test.go

```diff
@@ -0,0 +1,506 @@
+		require.ElementsMatch(t, expectedKey2Values, resp.GetResponse().GetValues())
+
+		ver = &types.Version{
+			BlockNum: 7,
+			TxNum:    0,
+		}
+		resp, err = s.GetNextValues(t, "db1", "key2", "alice", ver)
+		require.NoError(t, err)
+		require.Nil(t, resp.GetResponse().GetValues())
+	})
+
+	t.Run("Get previous values", func(t *testing.T) {
+		expectedKey2Values := []*types.ValueWithMetadata{
+			{
+				Value: []byte("key2value1"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 4,
+						TxNum:    0,
+					},
+					AccessControl: &types.AccessControl{
+						ReadWriteUsers: map[string]bool{"alice": true},
+					},
+				},
+			},
+			{
+				Value: []byte("key2value2"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 6,
+						TxNum:    0,
+					},
+				},
+			},
+		}
+
+		ver := &types.Version{
+			BlockNum: 7,
+			TxNum:    0,
+		}
+		resp, err := s.GetPreviousValues(t, "db1", "key2", "alice", ver)
+		require.NoError(t, err)
+		require.ElementsMatch(t, expectedKey2Values, resp.GetResponse().GetValues())
+
+		ver = &types.Version{
+			BlockNum: 4,
+			TxNum:    0,
+		}
+		resp, err = s.GetPreviousValues(t, "db1", "key2", "alice", ver)
+		require.NoError(t, err)
+		require.Nil(t, resp.GetResponse().GetValues())
+	})
+
+	t.Run("Get Most Recent Value at or Below", func(t *testing.T) {
+		expectedKey3Values := []*types.ValueWithMetadata{
+			{
+				Value: []byte("key3value2"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 7,
+						TxNum:    0,
+					},
+				},
+			},
+		}
+
+		ver := &types.Version{
+			BlockNum: 10,
+			TxNum:    7,
+		}
+		resp, err := s.GetMostRecentValueAtOrBelow(t, "db2", "key3", "bob", ver)
+		require.NoError(t, err)
+		require.ElementsMatch(t, expectedKey3Values, resp.GetResponse().GetValues())
+
+		ver = &types.Version{
+			BlockNum: 7,
+			TxNum:    0,
+		}
+		resp, err = s.GetMostRecentValueAtOrBelow(t, "db2", "key3", "bob", ver)
+		require.NoError(t, err)
+		require.ElementsMatch(t, expectedKey3Values, resp.GetResponse().GetValues())
+
+		expectedKey3Values = []*types.ValueWithMetadata{
+			{
+				Value: []byte("key3value1"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 4,
+						TxNum:    0,
+					},
+				},
+			},
+		}
+
+		ver = &types.Version{
+			BlockNum: 6,
+			TxNum:    0,
+		}
+		resp, err = s.GetMostRecentValueAtOrBelow(t, "db2", "key3", "bob", ver)
+		require.NoError(t, err)
+		require.ElementsMatch(t, expectedKey3Values, resp.GetResponse().GetValues())
+	})
+}
+
+// 1. Store key1 with open access to all users in database 1
+// 2. Store key2 with read-write access only for alice in database 1
+// 3. Store key3 with open acess to all users in database 2
+// 4. Bob reads key1 and update the same in database 1
+// 5. Alice reads key1 and key2, update key2 while deleting key1 in database 1
+// 6. Alice reads key2 from database 1 and key3 from database 2 and updates key2 and key3
```

Comment:
I am not sure this is enough data to cover all the use cases... maybe use a few of additional keys to create gaps between the versions?
Also, what happens when a key is written several time, and then deleted, and then added again?

Reply:
Good question. Though we have a similar scenario in the unit-test, there is a bug in the production code. Fixed it in #399